lib/Sema/CSApply.cpp

```diff
@@ -5399,10 +5399,16 @@ Expr *ExprRewriter::coerceCallArguments(Expr *arg, AnyFunctionType *funcType,
 
       // Otherwise, create a call of the default argument generator.
       } else {
-        defArg =
-            new (tc.Context) DefaultArgumentExpr(callee, paramIdx,
-                                                 arg->getStartLoc(),
-                                                 param.getParameterType());
+        const ParamDecl *defaultParam = getParameterAt(callee.getDecl(),
+                                                       newArgs.size());
+        defArg = defaultParam->getDefaultValue();
```

Comment:
I'm not sure we want to be inlining regular default arguments into the resulting AST, as it may cause problems for things like autoclosures, which get contextualized to the parameter's context rather than the call-site's context. In addition, I don't believe this will work in the general case for default arguments in other modules, as we currently don't deserialize an expression for them.

For example:

```swift
// Module A
public struct S {
  public subscript(_ x: Int = 0) -> Int { x }
}

// Module B
import A

func bar() {
  _ = \S.[]
}
```

When coercing the call arguments for the key-path subscript, `defaultParam->getDefaultValue()` will be `nullptr`.

The way we usually handle default arguments is to emit a "default argument generator" per default argument in the same file they're declared in. Other modules that want to use the default arguments emit a call to this generator function, and use the result as the default value.

I think we'll want to take a similar approach for key-path subscript defaults, so we'll want to carry on using `DefaultArgumentExpr` to represent the argument in the type-checked AST, but modify SILGen to emit a call to the appropriate default argument generator when lowering the arguments for the subscript.

I'm not too familiar with SILGen, but I think you'll want to see if you can use `ArgEmitter` to emit the arguments for you, this should handle `DefaultArgumentExpr` as a "delayed argument", which you should be able to emit afterwards through `emitDelayedArguments`. Does this sound reasonable @slavapestov?

Reply:
Do you think it would be beneficial to add a test with two modules? Also, just read the part about `ArgEmitter`. I'll see if I can use that instead.

Reply:
@zoecarver If we don't already have a test for cross-module default args, it would definitely be worth adding one